src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1291,7 +1272,41 @@ public Boolean visitUndetVar(UndetVar t, Type s) {
             public Boolean visitErrorType(ErrorType t, Type s) {
                 return true;
             }
-        };
+        }
+
+        public NullabilityComparator nullabilityComparator = new NullabilityComparator();
+        public class NullabilityComparator extends TypeRelation {
+            BiFunction<Type, Type, Boolean> differentNullability;
+            Warner warner;
+
+            NullabilityComparator reset(BiFunction<Type, Type, Boolean> differentNullability) {
+                this.differentNullability = differentNullability;
+                if (this.warner == null || this.warner == noWarnings) {
+                    this.warner = !warnStack.isEmpty() ? warnStack.head : noWarnings;
+                }
+                return this;
+            }
+
+            public NullabilityComparator setWarner(Warner warner) {
+                this.warner = warner;
+                return this;
+            }
+
+            public NullabilityComparator clearWarner() {
+                this.warner = null;
+                return this;
+            }
+
+            @Override
+            public Boolean visitType(Type t, Type s) {
+                if (differentNullability.apply(t, s)) {
```

Comment:
Is this truly a recursive visitor? I'm wondering, because if you don't need the visitor capabilities, then you could just drop the use of lambdas which will make the code a bit cleaner. E.g. instead of:
```
nullabilityComparator.reset((t1, t2) -> !t1.sameNullabilityAs(t2))
                            .visit(s, t);
```
do:
```
if (!s.sameNullabilityAs(t)) {
     nullabilityComparator.warn(...);
}
```
In which case, maybe NullabilityComparator is just NullabilityWarner?

Reply:
yes the visitor is needed, I will be adding another iteration with more test cases, this was an unfinished iteration